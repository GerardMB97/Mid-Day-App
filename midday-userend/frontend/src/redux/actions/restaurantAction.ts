import axios from 'axios';
import restaurantActionTypes from './restaurantActionTypes';

export const loadCategories = () => {
  return async (dispatch: Function) => {
    const categories = await axios.get('http://localhost:5000/api/midday/restaurants/categories');

    dispatch({
      type: restaurantActionTypes.LOAD_CATEGORIES,
      categories
    });
  };
}
;
