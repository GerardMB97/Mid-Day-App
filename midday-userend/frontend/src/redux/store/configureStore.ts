import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import reduxImmutableStateInvariant from 'redux-immutable-state-invariant';
import rootReducer from '../reducers';
import initialState from './initialState';

function configureStore () {
  const composeEnhancers = compose;

  return createStore(
    rootReducer,
    initialState,
    composeEnhancers(applyMiddleware(reduxImmutableStateInvariant(), thunk))
  );
}

export default configureStore();
