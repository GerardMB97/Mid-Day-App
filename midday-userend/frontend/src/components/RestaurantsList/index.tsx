import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import {
  Text,
  StyleSheet,
  FlatList,
  ImageBackground,
  TouchableOpacity,
  View
} from 'react-native';
import { State } from '../../models/index';
import { AnyAction, bindActionCreators, Dispatch } from 'redux';

import { getCategoryRestaurants } from '../../redux/actions/restaurantAction';
import SearchBar from '../SearchBar';
import colors from '../../../colors';

const styles = StyleSheet.create({
  container: {
    marginTop: 50,
    flex: 1,
    alignItems: 'center',
    width: '100%',
    paddingLeft: 20,
    paddingRight: 20
  },
  list: {
    flexDirection: 'column',
    width: '90%'
  },
  listElement: {
    width: '100%',
    height: 150,
    borderRadius: 15,
    marginBottom: 15,
    borderColor: 'black',
    borderWidth: 1,
    overflow: 'hidden',
    position: 'relative'
  },
  nameContainer: {
    backgroundColor: colors.transparentGray,
    width: '100%',
    height: 60,
    borderBottomLeftRadius: 15,
    borderBottomRightRadius: 15,
    alignItems: 'center',
    position: 'absolute',
    top: 90,
    flexDirection: 'row',
    justifyContent: 'space-between',
    padding: 20
  },
  image: {
    width: '100%',
    height: '100%',
    position: 'absolute',
    top: 0
  },
  nav: {
    width: '100%',
    height: 150
  },
  badge: {
    borderRadius: 50,
    backgroundColor: colors.green,
    width: 40,
    height: 40,
    alignItems: 'center',
    justifyContent: 'center'
  },
  price: {
    fontSize: 12,
    color: 'black'
  }
});
function RestaurantsList ({ route, restaurants, actions, navigation }:any) {
  const [inputValue, setInputValue] = React.useState('');
  const { category } = route?.params;
  useEffect(() => {
    actions.getCategoryRestaurants(category);
  }, [category]);

  return (
    <View style={styles.container}>
    <SearchBar
    inputPlaceholder = 'A qué restaurante te apetece ir hoy?'
    setInputValue ={setInputValue}
    inputValue = {inputValue}
    action = 'restaurants'
    ></SearchBar>
    <FlatList
        style = {styles.list}
        numColumns={1}
        data ={inputValue.length ? restaurants.filteredRestaurants : restaurants.categoryRestaurants}
        keyExtractor = {(item => item.name)}
        renderItem={({ item }) =>
          <View style = {styles.listElement} >

            <ImageBackground source= {{ uri: item.image }} style = {styles.image} >
              <TouchableOpacity testID={item.name} style = {styles.nav} onPress={() => { navigation.navigate('RestaurantDetail', { _id: item._id }); }}>
                <View style={styles.nameContainer}>
                <Text>{item.name}</Text>
                <Text>{`${item.street}, ${item.number}`}</Text>
                <View style={styles.badge}>
                  <Text style={styles.price}>{item.menuprice} €</Text>
                </View>
                </View>
                </TouchableOpacity>
            </ImageBackground>

          </View>}
      >
      </FlatList>
      </View>
  );
}

function mapStateToProps ({ restaurants }:{restaurants:State['restaurants']}) {
  return { restaurants };
}

function mapDispatchToProps (dispatch: Dispatch<AnyAction>) {
  return { actions: bindActionCreators({ getCategoryRestaurants }, dispatch) };
}

export default connect(mapStateToProps, mapDispatchToProps)(RestaurantsList);
