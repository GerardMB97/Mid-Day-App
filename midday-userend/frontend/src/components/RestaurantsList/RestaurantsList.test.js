import React from 'react';
import { Provider } from 'react-redux';
import { render, fireEvent } from 'react-native-testing-library';
import configureStore from 'redux-mock-store';
import * as actions from '../../redux/actions/restaurantActions/restaurantAction';
import RestaurantsList from '../RestaurantsList';

jest.mock('../../redux/actions/restaurantActions/restaurantAction');
jest.mock('react-native-vector-icons/Ionicons', () => 'Icon');
jest.mock('@react-navigation/native');

describe('Given a component RestaurantsList', () => {
  const restaurants = {
    allRestaurants: [{ _id: '3' }],
    categoryRestaurants: [{ name: 'asian', _id: '1' }],
    filteredRestaurants: [{ name: 'hi', _id: '2' }]
  };
  jest.spyOn(actions, 'getCategoryRestaurants').mockReturnValue({ type: '' });
  const mockStore = configureStore();
  const store = mockStore({ restaurants });

  describe('When invoked', () => {
    test('Then it should render correctly', () => {
      const tree = render(<Provider store={store}><RestaurantsList route={{ params: 'hi' }} navigation={{ isFocused: jest.fn() }}></RestaurantsList></Provider>);

      expect(tree).toMatchSnapshot();
    });
  });
  describe('When has inputValue.length ', () => {
    test('Then it should render filteredRestaurants', () => {
      const realUseState = React.useState;
      const stubInitialState = 'asiatica';

      jest
        .spyOn(React, 'useState')
        .mockImplementationOnce(() => realUseState(stubInitialState));
      const tree = render(<Provider store={store}><RestaurantsList route={{ params: 'hi' }}></RestaurantsList></Provider>);

      expect(tree).toMatchSnapshot();
    });
  });
  describe('When a restaurant is clicked', () => {
    test('Then it should navigate', () => {
      const navigation = {
        navigate: jest.fn()
      };

      const { getByTestId } = render(<Provider store={store}><RestaurantsList route={{ params: 'hi' }} navigation = {navigation}/></Provider>);
      fireEvent.press(getByTestId('asian'));

      expect(navigation.navigate).toHaveBeenCalled();
    });
  });
});
