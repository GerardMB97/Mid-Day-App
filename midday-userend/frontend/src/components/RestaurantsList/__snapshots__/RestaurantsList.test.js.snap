// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a component RestaurantsList When has inputValue.length  Then it should render filteredRestaurants 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "marginTop": 50,
      "paddingLeft": 20,
      "paddingRight": 20,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "paddingLeft": 20,
        "paddingRight": 20,
        "width": "100%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="A qué restaurante te apetece ir hoy?"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "rgb(135, 175, 147)",
          "borderRadius": 10,
          "height": 60,
          "justifyContent": "center",
          "marginBottom": 20,
          "paddingLeft": 40,
          "width": "100%",
        }
      }
      testID="searchbar"
      underlineColorAndroid="transparent"
      value="asiatica"
    />
    <Icon
      name="search-outline"
      style={
        Object {
          "color": "black",
          "fontSize": 25,
          "left": "10%",
          "position": "absolute",
          "top": "20%",
        }
      }
    />
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "name": "hi",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flexDirection": "column",
        "width": "90%",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "borderColor": "black",
              "borderRadius": 15,
              "borderWidth": 1,
              "height": 150,
              "marginBottom": 15,
              "overflow": "hidden",
              "position": "relative",
              "width": "100%",
            }
          }
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              Object {
                "height": "100%",
                "position": "absolute",
                "top": 0,
                "width": "100%",
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": undefined,
                }
              }
              style={
                Array [
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  Object {
                    "height": "100%",
                    "width": "100%",
                  },
                  undefined,
                ]
              }
            />
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "height": 150,
                  "opacity": 1,
                  "width": "100%",
                }
              }
              testID="hi"
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "rgba(242, 242, 242, 0.8)",
                    "borderBottomLeftRadius": 15,
                    "borderBottomRightRadius": 15,
                    "flexDirection": "row",
                    "height": 60,
                    "justifyContent": "space-between",
                    "padding": 20,
                    "position": "absolute",
                    "top": 90,
                    "width": "100%",
                  }
                }
              >
                <Text>
                  hi
                </Text>
                <Text>
                  undefined, undefined
                </Text>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "rgb(135, 175, 147)",
                      "borderRadius": 50,
                      "height": 40,
                      "justifyContent": "center",
                      "width": 40,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "black",
                        "fontSize": 12,
                      }
                    }
                  >
                     €
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Given a component RestaurantsList When invoked Then it should render correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "marginTop": 50,
      "paddingLeft": 20,
      "paddingRight": 20,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "paddingLeft": 20,
        "paddingRight": 20,
        "width": "100%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="A qué restaurante te apetece ir hoy?"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "rgb(135, 175, 147)",
          "borderRadius": 10,
          "height": 60,
          "justifyContent": "center",
          "marginBottom": 20,
          "paddingLeft": 40,
          "width": "100%",
        }
      }
      testID="searchbar"
      underlineColorAndroid="transparent"
      value=""
    />
    <Icon
      name="search-outline"
      style={
        Object {
          "color": "black",
          "fontSize": 25,
          "left": "10%",
          "position": "absolute",
          "top": "20%",
        }
      }
    />
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "name": "asian",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flexDirection": "column",
        "width": "90%",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "borderColor": "black",
              "borderRadius": 15,
              "borderWidth": 1,
              "height": 150,
              "marginBottom": 15,
              "overflow": "hidden",
              "position": "relative",
              "width": "100%",
            }
          }
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              Object {
                "height": "100%",
                "position": "absolute",
                "top": 0,
                "width": "100%",
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": undefined,
                }
              }
              style={
                Array [
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  Object {
                    "height": "100%",
                    "width": "100%",
                  },
                  undefined,
                ]
              }
            />
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "height": 150,
                  "opacity": 1,
                  "width": "100%",
                }
              }
              testID="asian"
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "rgba(242, 242, 242, 0.8)",
                    "borderBottomLeftRadius": 15,
                    "borderBottomRightRadius": 15,
                    "flexDirection": "row",
                    "height": 60,
                    "justifyContent": "space-between",
                    "padding": 20,
                    "position": "absolute",
                    "top": 90,
                    "width": "100%",
                  }
                }
              >
                <Text>
                  asian
                </Text>
                <Text>
                  undefined, undefined
                </Text>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "rgb(135, 175, 147)",
                      "borderRadius": 50,
                      "height": 40,
                      "justifyContent": "center",
                      "width": 40,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "black",
                        "fontSize": 12,
                      }
                    }
                  >
                     €
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
